include/mbedtls/mps/writer.h

```diff
@@ -1,50 +1,63 @@
+ *
+ * For example, the provider might be the [D]TLS record layer, offering
+ * to protect and transport data in records of varying size (depending
+ * on the current configuration and the amount of data left in the current
+ * datagram, for example), while the consumer would be the handshake logic
+ * layer which needs to write handshake messages. The size of handshake
+ * messages are entirely independent of the size of records used to transport
+ * them, and the writer helps to both split large handshake messages across
+ * multiple records, and to pack multiple small handshake messages into
+ * a single record. This example will be elaborated upon in the next paragraph.
+ *
+ * Basic flow of operation:
+ * First, the provider feeds an outgoing data buffer to the writer, transferring
+ * it from 'providing' to 'consuming' mode; in the example, that would be record
+ * layer providing the plaintext buffer for the next outgoing record. The
+ * consumer subsequently fetches parts of the buffer and writes data to them,
+ * which might happen multiple times; in the example, the handshake logic
+ * layer might request and fill a buffer for each handshake message in the
+ * current outgoing flight, and these requests would be served from successive
+ * chunks in the same record plaintext buffer if size permits. Once the consumer
+ * is done, the provider revokes the writer's access to the data buffer,
+ * putting the writer back to providing mode, and processes the data provided
+ * in the outgoing data buffer; in the example, that would be the record layer
+ * encrypting the record and dispatching it to the underlying transport.
+ * Afterwards, the provider feeds another outgoing data buffer to the writer
+ * and the cycle starts again.
+ * In the event that a consumer's request cannot be fulfilled on the basis of
+ * the outgoing data buffer provided by the provider (in the example,
+ * the handshake layer might attempt to send a 4KB CRT chain but the current
+ * record size offers only 2KB), the writer transparently offers a temporary
+ * 'queue' buffer to hold the data to the consumer. The contents of this queue
+ * buffer will be gradually split among the next outgoing data buffers when
+ * the provider subsequently provides them; in the example, the CRT chain would
+ * be split amont multiple records when the record layer hands more plaintext
+ * buffers to the writer. The details of this process are left to the writer
+ * and are opaque both to the consumer and the provider.
+ *
+ * Abstract models:
+ * From the perspective of the consumer, the state of the writer is a
+ * potentially empty list of output buffers that the writer has provided
+ * to the consumer. New buffers can be requested through calls to
+ * mbedtls_writer_get(), while previously obtained output buffers can be
+ * marked processed through calls to mbedtls_writer_commit(), emptying the
+ * list of output buffers and invalidating them from the consumer's perspective.
```

Comment:
This paragraph could benefit from a description of the model from the provider's perspective.